File: 프로그래머스/2/12913. 땅따먹기/땅따먹기.py
from functools import reduce
def solution(land):
    return max(reduce(lambda a, c : 
         [
            c[0] + max(a[1], a[2], a[3]),  
            c[1] + max(a[0], a[2], a[3]),
            c[2] + max(a[0], a[1], a[3]),
            c[3] + max(a[0], a[1], a[2]),
        ],  land, [0, 0, 0, 0]))

File: 프로그래머스/2/12913. 땅따먹기/땅따먹기.java
import java.util.*;
import java.util.stream.*;
class Solution {
    int solution(int[][] land) {
        int answer = 0;
        int [][] dp = new int[land.length][4];
        for(int i = 0; i<4;i++){
            dp[0][i] = land[0][i]; 
        }
        
        for(int j = 1; j< land.length; j++){
            for(int i = 0; i<4;i++){
                int idx1 = (i + 1) % 4;
                int idx2 = (i + 2) % 4;
                int idx3 = (i + 3) % 4;
                dp[j][i] = land[j][i] + Math.max(dp[j-1][idx1], Math.max(dp[j-1][idx2], dp[j-1][idx3])); 
            }
        }

        Integer[] ans = Arrays.stream(dp[dp.length-1]).boxed().toArray(Integer[]::new);
        return Collections.max(new ArrayList<>(Arrays.asList(ans)));
    }
}